File: project_pyratio/recipes.py
# recipes.py
# A type representing DSP recipes including inputs, outputs, and time periods

from dataclasses import dataclass

from facilities import FacilityCategory
from products import ProductQuantity
from rates import Time


@dataclass(init=False, frozen=True, slots=True)
class Recipe:
    name: str
    # I'm using tuples here because I want recipes to be *strongly* immutable
    # All types that represent DSP game concepts should be immutable for safety
    outputs: tuple[ProductQuantity[int], ...]
    inputs: tuple[ProductQuantity[int], ...]
    period: Time
    made_in: FacilityCategory

    def __init__(
        self,
        name: str | None,
        outputs: list[ProductQuantity[int]],
        inputs: list[ProductQuantity[int]],
        period: Time,
        made_in: FacilityCategory
    ) -> None:
        if len(outputs) == 1:
            autoname = outputs[0].name

        # Recipes with no special name are always named after their sole output
        if name is None:
            name = autoname

        cleaned_name = name.strip()
        if cleaned_name == "":
            raise ValueError("recipe name empty or all whitespace")

        # At some point, I may want to allow empty input or output lists...
        # ...but for the time being, these checks exist to preserve my sanity
        if len(outputs) < 1:
            raise ValueError("recipe output collection is empty")
        if len(inputs) < 1:
            raise ValueError("recipe input collection is empty")

        # Let's enforce the invariant of making sure the tuples are sorted
        # The actual order doesn't deeply matter, thus the use of hash()
        # What we care about here is that the ordering will be *consistent*
        # This invariant is important for equality comparisons between recipes
        # Vectors passed into this constructor could have elements in any order
        # That order should not affect whether recipe instances compare equal
        tuple_outputs = tuple(sorted(outputs, key=hash))
        tuple_inputs = tuple(sorted(inputs, key=hash))

        object.__setattr__(self, "name", cleaned_name)
        object.__setattr__(self, "outputs", tuple_outputs)
        object.__setattr__(self, "inputs", tuple_inputs)
        object.__setattr__(self, "period", period)
        object.__setattr__(self, "made_in", made_in)

    @property
    def category(self) -> FacilityCategory:
        return self.made_in

    # Single line recipe printing
    def __repr__(self) -> str:
        parts: list[str] = []

        parts.append(self.name)
        parts.append(": ")

        parts.append("[")
        parts.append(", ".join(str(output) for output in self.outputs))
        parts.append("]")

        parts.append(" <<< ")
        parts.append(f"{self.period} ({self.category})")
        parts.append(" <<< ")

        parts.append("[")
        parts.append(", ".join(str(input_) for input_ in self.inputs))
        parts.append("]")

        return "".join(parts)

    # Multi-line recipe printing
    def __str__(self) -> str:
        RECIPE_INDENT: str = " " * 2
        parts: list[str] = []

        parts.append(self.name)
        parts.append(":\n")

        parts.append(f"{RECIPE_INDENT}[")
        parts.append(", ".join(str(output) for output in self.outputs))
        parts.append("]\n")

        parts.append(f"{RECIPE_INDENT * 2}^\n{RECIPE_INDENT * 2}^ ")
        parts.append(f"{self.period} ({self.category})")
        parts.append(f"\n{RECIPE_INDENT * 2}^\n")

        parts.append(f"{RECIPE_INDENT}[")
        parts.append(", ".join(str(input_) for input_ in self.inputs))
        parts.append("]")

        return "".join(parts)

File: julia_code/recipes.jl
# recipes.jl
# A type representing DSP recipes including inputs, outputs, and time periods


struct Recipe
	name::String
	# I'm using tuples here because I want recipes to be *strongly* immutable
	# All types that represent DSP game concepts should be immutable for safety
	outputs::Tuple{Vararg{ProductQuantity{Int}}}
	inputs::Tuple{Vararg{ProductQuantity{Int}}}
	period::Time
	madein::FacilityCategory
	
	function Recipe(name::AbstractString,
		outputs::AbstractVector{ProductQuantity{Int}},
		inputs::AbstractVector{ProductQuantity{Int}}, period::Time,
		madein::FacilityCategory)
		
		cleanedname = strip(name)
		isempty(cleanedname) && error("recipe name empty or all whitespace")
		
		# At some point, I may want to allow empty input or output lists...
		# ...but for the time being, these checks exist to preserve my sanity
		isempty(outputs) && error("recipe output collection is empty")
		isempty(inputs) && error("recipe input collection is empty")
		
		# Let's enforce the invariant of making sure the tuples are sorted
		# The actual order doesn't deeply matter, thus the use of hash()
		# What we care about here is that the ordering will be *consistent*
		# This invariant is important for equality comparisons between recipes
		# Vectors passed into this constructor could have elements in any order
		# That order should not affect whether recipe instances compare equal
		tupleoutputs = Tuple(sort(outputs, by = hash))
		tupleinputs = Tuple(sort(inputs, by = hash))
		
		new(string(cleanedname), tupleoutputs, tupleinputs, period, madein)
	end
end

function Recipe(outputs::AbstractVector{ProductQuantity{Int}},
	inputs::AbstractVector{ProductQuantity{Int}}, period::Time,
	madein::FacilityCategory)
	
	# Recipes with no special name are always named after their sole output
	autoname = name(only(outputs))
	
	Recipe(autoname, outputs, inputs, period, madein)
end

function Recipe(name::Nothing, outputs::AbstractVector{ProductQuantity{Int}},
	inputs::AbstractVector{ProductQuantity{Int}}, period::Time,
	madein::FacilityCategory)
	
	Recipe(outputs, inputs, period, madein)
end

name(r::Recipe) = r.name
outputs(r::Recipe) = r.outputs
inputs(r::Recipe) = r.inputs
period(r::Recipe) = r.period
category(r::Recipe) = r.madein

# Single line recipe printing
function Base.show(io::IO, r::Recipe)
	
	print(io, name(r))
	print(io, ": ")
	
	print(io, "[")
	join(io, (string(output) for output in outputs(r)), ", ")
	print(io, "]")
	
	print(io, " <<< ")
	print(io, period(r), " (", category(r), ")")
	print(io, " <<< ")
	
	print(io, "[")
	join(io, (string(input) for input in inputs(r)), ", ")
	print(io, "]")
end

# Multi-line recipe printing
function Base.show(io::IO, ::MIME"text/plain", r::Recipe)
	
	RECIPEINDENT = " " ^ 2
	
	print(io, name(r))
	print(io, ":\n")
	
	print(io, RECIPEINDENT, "[")
	join(io, (string(output) for output in outputs(r)), ", ")
	print(io, "]\n")
	
	print(io, RECIPEINDENT ^ 2, "^\n", RECIPEINDENT ^ 2, "^ ")
	print(io, period(r), " (", category(r), ")")
	print(io, "\n", RECIPEINDENT ^ 2, "^\n")
	
	print(io, RECIPEINDENT, "[")
	join(io, (string(input) for input in inputs(r)), ", ")
	print(io, "]")
end

